File: app/src/main/java/com/tama/command/Command.java
package com.tama.command;

import com.tama.core.Pet;
import com.tama.core.World;

/**
 * An abtraction for any single action that a pet can execute
 */
public abstract class Command
{
    interface Update
    {
        void update(Pet pet, World world);
    }
    public Update updater = this::start;

    public int priority = 0;
    public CommandState state = CommandState.doing;

    protected abstract void start(Pet pet, World world);
    protected abstract void doing(Pet pet, World world);
}

File: app/src/main/java/com/tama/command/Command.kt
package com.tama.command

import com.tama.core.DisplayAdapter
import com.tama.core.World
import com.tama.thing.Pet
import java.io.Serializable

/**
 * An abtraction for any single action that a pet can execute
 */
abstract class Command : Serializable
{
    public var update: (pet: Pet, world: World) -> Unit = ::start
    public var state = CommandState.ready
    var actor: Pet? = null;

    protected open fun start(pet: Pet, world: World)
    {
        actor = pet;
        state = CommandState.doing;
        update = ::doing;
    }

    protected abstract fun doing(pet: Pet, world: World);

    open fun isReplaceable(): Boolean
    {
        return state == CommandState.ready;
    }

    open fun hardCancel()
    {
        state = CommandState.failed;
    }

    abstract fun draw(d: DisplayAdapter);
}